### PR TITLE
Allow arbitrary fluxes across immersed boundaries and test this with `AdvectiveForcing`

### PR DESCRIPTION
**Pull request by tomchor**
This PR makes it possible for users to enforce both open and closed boundaries for `AdvectiveForcing` for both immersed and domain boundaries. This is done by removing methods that enforced zero-flux through immersed boundaries. I'm also adding a test for this and a validation script which produces the following animations for open and closed boundaries, respectively:

https://github.com/user-attachments/assets/457feb2e-f5d6-4f69-a02c-bf6ee2970b7c

https://github.com/user-attachments/assets/8dc8844c-6627-484f-9c28-95db7653102e

I also fixed a few errors/typos in other validation scripts.

~Importantly, even after this PR, if a user passes a `Number` to `AdvectiveForcing` the behavior at domain and immersed boundaries remains different. Therefore this does *not* close https://github.com/CliMA/Oceananigans.jl/issues/4812, since a `Field` with proper boundary conditions is required to unify behavior across different boundaries.~

### CODE REVIEW

File: src/Forcings/advective_forcing.jl [hunk 1]
```diff
@@ -1,10 +1,18 @@
...
 using Oceananigans.Utils: sum_of_velocities
+using Oceananigans.BoundaryConditions: OpenBoundaryCondition, FieldBoundaryConditions
+using Oceananigans.ImmersedBoundaries: ImmersedBoundaryCondition
 using Adapt
 
-maybe_constant_field(u) = u
-maybe_constant_field(u::Number) = ConstantField(u)
+maybe_field(u, args...) = u
+
+function maybe_field(u::Number, loc, grid, u_bcs)
```
Comment (glwagner) [inline src/Forcings/advective_forcing.jl]: is this different from the function `field`?

Comment (tomchor) [reply src/Forcings/advective_forcing.jl]: It is if `u` isn't a Number. (For example and array or a `ConstantField`.)
There's a slight difference for numbers as well since `u_bcs` isn't ` FieldBoundaryConditions`. (I'm trying to keep the interface simple and avoid having the user pass a whole set of `FieldBoundaryConditions` when creating a simple slip velocity forcing.

Comment (glwagner) [reply src/Forcings/advective_forcing.jl]: Hmm... is it all simpler if we just convert `Number` to `Field` and then enforce boundary conditions correctly? (Sorry if I am misundersatnding something)

The other question is whether we can generalize `field` to encompass this case.

Comment (tomchor) [reply src/Forcings/advective_forcing.jl]: > Hmm... is it all simpler if we just convert `Number` to `Field` and then enforce boundary conditions correctly? (Sorry if I am misundersatnding something)

Don't you need to pass boundary conditions to the `Field` when creating it? Because if so, that's what this code does (I simplified it a bit more if you wanna take another look).

> The other question is whether we can generalize `field` to encompass this case.

That'd definitely simplify things in this particular instance. But I think we'd need to create a `Field(f::AbstractField, args...) = f` method. Would such method ignore boundary conditions, indices, etc. that are passed as `args`? Or would it "copy" `f` and change the it according to `args`? I don't think any of these options is great to be honest. What do you think?

Comment (glwagner) [reply src/Forcings/advective_forcing.jl]: > Don't you need to pass boundary conditions to the Field when creating it? Because if so, that's what this code does (I simplified it a bit more if you wanna take another look).

Yes, `Field` need boundary conditions if they are not default. For velocities, I would:

1. Use default velocity boundary conditions if users give a Number (eg impenetrable)
2. Also support explicitly providing `Field` which bundle boundary conditions

**Comment by tomchor** [comment]
Re-centering and updating the discussion around face boundary conditions for `AdvectiveForcing` here. Per https://github.com/CliMA/Oceananigans.jl/issues/4998#issuecomment-3592754986, @glwagner suggested that the BCs for `AdvectiveForcing` be imposed by masking the immersed values to the appropriate value.

This by itself doesn't seem to work. For example, in the example below, `AdvectiveForcing.w` is equal to -0.02 everywhere, including in the immersed solid and domain faces. I'd expect both cases to result in the tracer passing through the boundaries, but instead the tracer accumulates in the immersed boundary:

https://github.com/user-attachments/assets/72b80ba5-ffda-4eac-b5b4-c76a2467cacd

@glwagner did I miss something about your suggestion?

**Comment by glwagner** [comment]
> Re-centering and updating the discussion around face boundary conditions for `AdvectiveForcing` here. Per [#4998 (comment)](https://github.com/CliMA/Oceananigans.jl/issues/4998#issuecomment-3592754986), @glwagner suggested that the BCs for `AdvectiveForcing` be imposed by masking the immersed values to the appropriate value.
> 
> This by itself doesn't seem to work. For example, in the example below, `AdvectiveForcing.w` is equal to -0.02 everywhere, including in the immersed solid and domain faces. I'd expect both cases to result in the tracer passing through the boundaries, but instead the tracer accumulates in the immersed boundary:
> 
>  nonhydrostatic_tracer_circle_2d_compare.mp4 
> @glwagner did I miss something about your suggestion?

You didn't miss anything. This reveals that there is more work to do. Try commenting out these specializations for immersed boundary grid:

https://github.com/CliMA/Oceananigans.jl/blob/d23257c9c9b0bce70923b2e5cdb809885188c02b/src/Advection/immersed_advective_fluxes.jl#L57-L59

these seem to hardcode impenetrability for an immersed boundary grid.

@simone-silvestri may want to comment on this. I think we have discussed about how the conditional advection operators are redundant with masking. If users want to have open immersed boundaries, we will need to stop using conditional advection operators that assume impenetrability on immersed boundaries.

**Comment by simone-silvestri** [comment]
I guess we could just remove these. The flux on the immersed boundaries will be zeroed by masking the velocities.

**Comment by tomchor** [comment]
This does seem to work. Let's see if removing this doesn't break any tests.

File: src/Advection/immersed_advective_fluxes.jl [hunk 2]
```diff
@@ -54,10 +54,6 @@ end
 @inline _advective_momentum_flux_Vw(i, j, k, ibg::IBG, scheme, V, w) = conditional_flux_cff(i, j, k, ibg, zero(ibg), advective_momentum_flux_Vw(i, j, k, ibg, scheme, V, w))
 @inline _advective_momentum_flux_Ww(i, j, k, ibg::IBG, scheme, W, w) = conditional_flux_ccc(i, j, k, ibg, zero(ibg), advective_momentum_flux_Ww(i, j, k, ibg, scheme, W, w))
 
-@inline _advective_tracer_flux_x(i, j, k, ibg::IBG, scheme, U, c) = conditional_flux_fcc(i, j, k, ibg, zero(ibg), advective_tracer_flux_x(i, j, k, ibg, scheme, U, c))
-@inline _advective_tracer_flux_y(i, j, k, ibg::IBG, scheme, V, c) = conditional_flux_cfc(i, j, k, ibg, zero(ibg), advective_tracer_flux_y(i, j, k, ibg, scheme, V, c))
-@inline _advective_tracer_flux_z(i, j, k, ibg::IBG, scheme, W, c) = conditional_flux_ccf(i, j, k, ibg, zero(ibg), advective_tracer_flux_z(i, j, k, ibg, scheme, W, c))
```
Comment (glwagner) [inline src/Advection/immersed_advective_fluxes.jl:59]: I would suggest also removing the momentum fluxes for consistency, although it is not necessary for AdvectiveForcing

**Comment by simone-silvestri** [comment]
I think the momentum fluxes might require a bit more work, since momentum fluxes always require interpolation of both advecting and advected velocities, and we do not have three-dimensional interpolation operators yet, so the velocity interpolated at the face might not be zero, leading to a non-zero flux across the immersed boundary.

File: src/Advection/immersed_advective_fluxes.jl (hunk 2 above)
Comment (tomchor) [reply src/Advection/immersed_advective_fluxes.jl:59]: Based on @simone-silvestri 's comment maybe we should leave this for another PR then?

File: src/Forcings/advective_forcing.jl [hunk 3]
```diff
@@ -1,9 +1,26 @@
...
+maybe_field(u::AbstractField, args...) = u
+
+function maybe_field(u::Number, loc, grid, u_bcs, open_boundaries)
+    immersed_bcs = ImmersedBoundaryCondition(; u_bcs...)
+    bcs = FieldBoundaryConditions(grid, loc; u_bcs..., immersed = immersed_bcs)
+    u_field = Field(loc, grid; boundary_conditions = bcs)
+    set!(u_field, u)
+    fill_halo_regions!(u_field)
+    if open_boundaries
+        mask_immersed_field!(u_field, u)
```
Comment (simone-silvestri) [inline src/Forcings/advective_forcing.jl]: Wouldn't this fill `u` over all the immersed boundaries?

Comment (tomchor) [reply src/Forcings/advective_forcing.jl]: Yes, but since the interior of the immersed solid isn't used in any calculation I think that's okay, no?

Comment (simone-silvestri) [reply src/Forcings/advective_forcing.jl]: I mean, if there is a complex boundary, all the boundaries end up advecting? Both west-facing and east-facing, for example. Isn't this a bit too much of an edge case to enforce with just a single keyword argument?

Comment (glwagner) [reply src/Forcings/advective_forcing.jl]: There's only advection across faces that are normal to the velocity component

Comment (glwagner) [reply src/Forcings/advective_forcing.jl]: I don't think we should use a keyword argument though, or anything specific to `AdvectiveForcing` at all. I think we should use regular boundary conditions and require users to build fields with those boundary conditions. We then should modify the `Field` constructor so that, if users have requested open immersed boundary conditions, the immersed region is masked.

For this to work generally we will have to fill every time step. But we can support constant ones to start.

Comment (glwagner) [reply src/Forcings/advective_forcing.jl]: happy to help with this @tomchor if what I said isn't very clear. The first thing to fix is the issue you raised that Face ImmersedBoundaryCondition are not respected.

This also addresses @simone-silvestri's point because the ImmersedBoundaryCondition is applied to specific facets of the cells.

Comment (tomchor) [reply src/Forcings/advective_forcing.jl]: I leaned towards having a boolean keyword argument specifically because `ImmersedBoundaryCondition` isn't respected. I think otherwise the interface becomes an little more unclear. It feels odd to me if a user passes complete `FieldBoundaryConditions` and then the velocities appear masked. There's still the option for the user to pass a `Field` with proper boundary conditions by the way.

> The first thing to fix is the issue you raised that Face ImmersedBoundaryCondition are not respected.

This was already fixed with your suggestion! Now we only gotta discuss the interface.

**Comment by tomchor** [comment]
For the record, as it is, this code has the desired effect (which is also tested in `test_forcings.jl`). This is the result of a simulation with open boundaries:

https://github.com/user-attachments/assets/a51dfe70-be10-407e-9190-d1b9e01f68ab

And the result for closed boundaries:

https://github.com/user-attachments/assets/17155ebd-df6f-4220-9abc-80ff584ec104

File: src/Forcings/advective_forcing.jl (hunk 3 above)
Comment (glwagner) [reply src/Forcings/advective_forcing.jl]: The problem with this solution is that it is not general. We want to have the same interface for boundary conditions for AdvectiveForcing and resolved velocity fields. I don't think we want a special interface just for forcing.

Comment (simone-silvestri) [reply src/Forcings/advective_forcing.jl]: I agree that we should probably solve this issue with a PR that modifies the field constructor (and mask immersed field) rather than advective forcing.

Comment (tomchor) [reply src/Forcings/advective_forcing.jl]: Yeah the current solution is not general. I do agree with @simone-silvestri that if we want to have something general, then I think this should be another PR dedicated to modifying the field constructor.

File: src/Forcings/advective_forcing.jl [hunk 4]
```diff
@@ -38,17 +63,38 @@ R = mean_particle_radius
...
+
+    if open_boundaries
+        u_bc = OpenBoundaryCondition(u)
+        v_bc = OpenBoundaryCondition(v)
+        w_bc = OpenBoundaryCondition(w)
+    else
+        u_bc = OpenBoundaryCondition(nothing)
+        v_bc = OpenBoundaryCondition(nothing)
+        w_bc = OpenBoundaryCondition(nothing)
+    end
```
Comment (glwagner) [inline src/Forcings/advective_forcing.jl]: what will happen if the user provides a field that has open boundary conditions embedded into it? Will this code work correctly?

File: test/test_forcings.jl [hunk 5]
```diff
@@ -252,6 +252,69 @@ function seven_forcings(arch)
...
+        return model
+    end
+
+    # Create models
+    w_settle = -0.01
+    regular_model = build_settling_model(regular_grid, w_settle)
+    immersed_model = build_settling_model(immersed_grid, w_settle)
+
+    regular_initial_integral = Integral(regular_model.tracers.c) |> Field |> deepcopy
+    immersed_initial_integral = Integral(immersed_model.tracers.c) |> Field |> deepcopy
```
Comment (glwagner) [inline test/test_forcings.jl]: since the `Integral` is used later, this will be easier to understand if the `Integral` fields are persisted and then used later, and the deepcopy is used to save the initial state which is compared to the later fields.

File: src/Forcings/advective_forcing.jl (hunk 4 above)
Comment (tomchor) [reply src/Forcings/advective_forcing.jl]: If the user provides a field then the field is used as-is.

File: test/test_forcings.jl (hunk 5 above)
Comment (tomchor) [reply test/test_forcings.jl]: Done!

File: src/Forcings/advective_forcing.jl (hunk 4 above)
Comment (glwagner) [reply src/Forcings/advective_forcing.jl]: ah ok great. So we can perhaps revert these changes to the `AdvectiveForcing` constructor and keep the contributions of this PR to the changes to the advection operators (a very significant contribution)! Then we can work on the user interface in a subsequent PR so that users don't have to manually call `mask_immersed_field!` and can use the boundary condition interface instead. What do you think?

Comment (glwagner) [reply src/Forcings/advective_forcing.jl]: If I understand correctly what is needed now is something like

```julia
w = ZFaceField(grid)
parent(w) .= 1
```

then `w` will work if used in `AdvectiveForcing`?

Comment (simone-silvestri) [reply src/Forcings/advective_forcing.jl]: I think so. In this case the tracer will pass through immersed boundaries since we have eliminated the `_advective_tracer_flux_x` methods.

Comment (glwagner) [reply src/Forcings/advective_forcing.jl]: Honestly the above is probably simpler than building an ImmersedBoundaryCondition, if your goal is to have a constant velocity

Comment (glwagner) [reply src/Forcings/advective_forcing.jl]: the boundary condition would be needed for time-dependent velocities (and possibly useful for spatially-varying as well)

Comment (tomchor) [reply src/Forcings/advective_forcing.jl]: > If I understand correctly what is needed now is something like
> 
> ```julia
> w = ZFaceField(grid)
> parent(w) .= 1
> ```
> 
> then `w` will work if used in `AdvectiveForcing`?

Yes

Comment (tomchor) [reply src/Forcings/advective_forcing.jl]: > ah ok great. So we can perhaps revert these changes to the `AdvectiveForcing` constructor and keep the contributions of this PR to the changes to the advection operators (a very significant contribution)! Then we can work on the user interface in a subsequent PR so that users don't have to manually call `mask_immersed_field!` and can use the boundary condition interface instead. What do you think?

Yeah, that sounds good to me!

File: docs/src/models/forcing_functions.md [hunk 6]
```diff
@@ -294,20 +294,22 @@ As an example, consider a model for sediment settling at a constant rate:
 ```jldoctest
 using Oceananigans
 
+grid = RectilinearGrid(size=(1, 1, 1), extent=(1, 1, 1))
+
```
Comment (tomchor) [inline docs/src/models/forcing_functions.md]:
```suggestion
```

File: docs/src/models/forcing_functions.md [hunk 7]
```diff
@@ -294,20 +294,22 @@ As an example, consider a model for sediment settling at a constant rate:
...
 
-sinking = AdvectiveForcing(w=w_sediment)
+sinking = AdvectiveForcing(w=w_sediment; grid)
 
 # output
 AdvectiveForcing:
 ├── u: ZeroField{Int64}
 ├── v: ZeroField{Int64}
-└── w: ConstantField(-0.00352102)
+└── w: 1×1×2 Field{Center, Center, Face} on RectilinearGrid on CPU
```
Comment (tomchor) [inline docs/src/models/forcing_functions.md]:
```suggestion
└── w: ConstantField(-0.00352102)
```

**Comment by tomchor** [comment]
@glwagner @simone-silvestri if tests pass this should be ready (they passed locally for me).

File: src/Advection/immersed_advective_fluxes.jl (hunk 2 above)
Comment (glwagner) [inline src/Advection/immersed_advective_fluxes.jl:59]: can we also delete the momentum flux functions? I don't want to risk inconsistency between momentum and tracers.

File: test/test_forcings.jl [hunk 8]
```diff
@@ -252,6 +253,88 @@ function seven_forcings(arch)
...
+function test_settling_tracer_comparison(arch; open_bottom=true)
+    """
+    Test that compares settling tracer simulations on regular vs immersed boundary grids.
+    Both should conserve tracer mass and have similar maximum values.
+    """
+
+    Nz = 16
+    Lz = 1
+
+    regular_grid = RectilinearGrid(arch, topology = (Flat, Flat, Bounded), size = (Nz,), z = (-Lz, 0))
```
Comment (glwagner) [inline test/test_forcings.jl]:
```suggestion
    regular_grid = RectilinearGrid(arch, topology = (Flat, Flat, Bounded), size = tuple(Nz), z = (-Lz, 0))
```

Comment (glwagner) [reply test/test_forcings.jl]: does `size=Nz` work? that would be preferred.

File: validation/advective_forcing/settling_tracer_blob.jl [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+using NCDatasets
+using CairoMakie
+
+function build_grid(use_immersed_grid::Bool;
+    Nx = 16,
+    Nz = 16,
+    Lx = 1,
+    Lz = 1,
+    )
```
Comment (glwagner) [inline validation/advective_forcing/settling_tracer_blob.jl:16]: hanging `)` violates what we put in AGENTS.md, but its ok here bc its just a validation case. just FYI

Comment (tomchor) [reply validation/advective_forcing/settling_tracer_blob.jl:16]: Changed

File: test/test_forcings.jl (hunk 8 above)
Comment (tomchor) [reply test/test_forcings.jl]: Yup, changed

File: src/Advection/immersed_advective_fluxes.jl (hunk 2 above)
Comment (tomchor) [reply src/Advection/immersed_advective_fluxes.jl:59]: Let's see if tests pass with this change

Comment (tomchor) [reply src/Advection/immersed_advective_fluxes.jl:59]: Tests pass! @glwagner  I think this should be gtg

File: test/test_forcings.jl [hunk 10]
```diff
@@ -252,6 +253,88 @@ function seven_forcings(arch)
...
+    w_settle = -0.01
+    regular_model = build_settling_model(regular_grid, w_settle)
+    immersed_model = build_settling_model(immersed_grid, w_settle)
+
+    ∫c_regular = Integral(regular_model.tracers.c) |> Field
+    ∫c_immersed = Integral(immersed_model.tracers.c) |> Field
+
+    regular_initial_integral = ∫c_regular |> deepcopy
+    immersed_initial_integral = ∫c_immersed |> deepcopy
+    @test regular_initial_integral[] == immersed_initial_integral[]
```
Comment (glwagner) [inline test/test_forcings.jl:307]: this incurs a scalar operation on GPU, but I guess we are globally allowing scalar ops, so its fine 😂

**Comment by simone-silvestri** [comment]
I was looking at this PR again and I still think removing the momentum flux zeroing on immersed boundaries might still be wrong because of the fact that we do not have boundary-aware interpolation. Is this tested? How was it solved?

**Comment by tomchor** [comment]
> I was looking at this PR again and I still think removing the momentum flux zeroing on immersed boundaries might still be wrong because of the fact that we do not have boundary-aware interpolation. Is this tested? How was it solved?

Is was tested with the added validation script and the added test, but that's about it.

**Comment by simone-silvestri** [comment]
Would it be possible to add a test for all the advection schemes that should that independently of the internal solution, when computing the momentum flux on peripheral nodes it is zero?

**Comment by tomchor** [comment]
@simone-silvestri added https://github.com/CliMA/Oceananigans.jl/pull/5402

Can you double check if that's what you meant?
